src/ILLink.Shared/XmlAttributeParser.cs

```diff
@@ -0,0 +1,317 @@
+			public string Name;
+			public ParameterNode (XPathNavigator nav) : base (nav)
+			{
+				Name = GetName (nav);
+			}
+		}
+
+		public partial record MethodNode : TypeMemberNode
+		{
+			public List<ParameterNode> Parameters;
+			public List<AttributeNode> ReturnAttributes;
+			public MethodNode (XPathNavigator nav) : base (nav)
+			{
+				Parameters = new List<ParameterNode> ();
+				ReturnAttributes = new List<AttributeNode> ();
+				foreach (XPathNavigator parameterNav in nav.SelectChildren (ParameterElementName, "")) {
+					Parameters.Add (new ParameterNode (parameterNav));
+				}
+				foreach (XPathNavigator returnNav in nav.SelectChildren (ReturnElementName, "")) {
+					foreach (XPathNavigator attributeNav in returnNav.SelectChildren (AttributeElementName, "")) {
+						ReturnAttributes.Add (new AttributeNode (attributeNav));
+					}
+				}
+			}
+		}
+
+		public abstract record TypeBaseNode : FeatureSwitchedNode
+		{
+			public List<TypeMemberNode> Events;
+			public List<TypeMemberNode> Fields;
+			public List<TypeMemberNode> Properties;
+			public List<MethodNode> Methods;
+			public List<NestedTypeNode> Types;
+			public TypeBaseNode (XPathNavigator nav) : base (nav)
+			{
+				Methods = new List<MethodNode> ();
+				Properties = new List<TypeMemberNode> ();
+				Fields = new List<TypeMemberNode> ();
+				Events = new List<TypeMemberNode> ();
+				foreach (XPathNavigator methodNav in nav.SelectChildren (MethodElementName, "")) {
+					Methods.Add (new MethodNode (methodNav));
+				}
+				foreach (XPathNavigator propertyNav in nav.SelectChildren (PropertyElementName, "")) {
+					Properties.Add (new TypeMemberNode (propertyNav));
+				}
+				foreach (XPathNavigator eventNav in nav.SelectChildren (EventElementName, "")) {
+					Events.Add (new TypeMemberNode (eventNav));
+				}
+				foreach (XPathNavigator fieldNav in nav.SelectChildren (FieldElementName, "")) {
+					Fields.Add (new TypeMemberNode (fieldNav));
+				}
+				Types = new List<NestedTypeNode> ();
+				foreach (XPathNavigator nestedTypeNav in nav.SelectChildren (TypeElementName, "")) {
+					Types.Add (new NestedTypeNode (nestedTypeNav));
+				}
+			}
+		}
+		public partial record NestedTypeNode : TypeBaseNode
+		{
+			public string Name;
+			public NestedTypeNode (XPathNavigator nav) : base (nav)
+			{
+				Name = GetName (nav);
+			}
+		}
+
+		public partial record TypeNode : TypeBaseNode, IRootNode
+		{
+			public string FullName;
+			public TypeNode (XPathNavigator nav) : base (nav)
+			{
+				FullName = GetFullName (nav);
+			}
+		}
+
+		public record AssemblyNode : FeatureSwitchedNode, IRootNode
+		{
+			public List<TypeNode> Types;
+			public string FullName;
+			public AssemblyNode (XPathNavigator nav) : base (nav)
+			{
+				FullName = GetFullName (nav);
+				Types = new List<TypeNode> ();
+				foreach (XPathNavigator typeNav in nav.SelectChildren (TypeElementName, "")) {
+					Types.Add (new TypeNode (typeNav));
+				}
+			}
+		}
+
+		public interface IRootNode
```

Comment:
Nit: In XML context "root" is the one element at the root of the document. If I read this correctly, this interface is not it - instead it's the "top-level" element in a way (the root is always "linker", right?).

Reply:
Good point, I updated the name to `ITopLevelNode`